spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md

```diff
@@ -521,53 +517,67 @@ function chanUpgradeTry(
             currentChannel.upgradeSequence = counterpartyUpgradeSequence - 1
         }
 
-        initUpgradeChannelHandshake(portIdentifier, channelIdentifier, upgradeFields, counterpartyUpgrade.timeout)
-    } else if currentChannel.state == INITUPGRADE {
-        existingUpgrade = publicStore.get(channelUpgradePath)
-        abortTransactionUnless(existingUpgrade.fields == upgradeFields)
+        initUpgradeChannelHandshake(portIdentifier, channelIdentifier, upgradeFields)
+    } else {
+        // we must use the existing upgrade fields
+        upgradeFields = existingUpgrade.fields
     }
 
+    abortTransactionUnless(isCompatible(upgradeFields, counterpartyUpgradeFields))
+
     // get counterpartyHops for given connection
     connection = getConnection(currentChannel.connectionIdentifier)
     counterpartyHops = getCounterpartyHops(connection)
 
     // construct counterpartyChannel from existing information and provided
     // counterpartyUpgradeSequence
     counterpartyChannel = ChannelEnd{
-        state: INITUPGRADE,
+        state: OPEN,
         ordering: currentChannel.ordering,
         counterpartyPortIdentifier: portIdentifier,
         counterpartyChannelIdentifier: channelIdentifier,
         connectionHops: counterpartyHops,
         version: currentChannel.version,
         sequence: counterpartyUpgradeSequence,
-        flushStatus: NOTINFLUSH,
     }
 
-    // call startFlushUpgrade handshake to move channel from INITUPGRADE to TRYUPGRADE and start flushing
-    // upgrade is blocked on this channelEnd from progressing until flush completes on both ends
-    startFlushUpgradeHandshake(portIdentifier, channelIdentifier, upgradeFields, counterpartyChannel, counterpartyUpgrade, TRYUPGRADE, proofChannel, proofUpgrade, proofHeight)
+    // verify proofs of counterparty state
+    abortTransactionUnless(verifyChannelState(connection, proofHeight, proofChannel, currentChannel.counterpartyPortIdentifier, currentChannel.counterpartyChannelIdentifier, counterpartyChannel))
+    abortTransactionUnless(verifyChannelUpgrade(connection, proofHeight, proofUpgrade, currentChannel.counterpartyPortIdentifier, currentChannel.counterpartyChannelIdentifier, counterpartyUpgrade))
+
+    // if the counterparty sequence is less than the current sequence, then either the counterparty chain is out-of-sync or
+    // the message is out-of-sync and we write an error receipt with our sequence - 1 so that the counterparty can update
+    // their sequence as well.
+    if counterpartyUpgradeSequence < channel.sequence {     
+        errorReceipt = ErrorReceipt{
+        channel.sequence - 1,
+        "sequence out of sync", // constant string changable by implementation
+        }
+        provableStore.set(channelUpgradeErrorPath(portIdentifier, channelIdentifier), errorReceipt)
+        return
```

Comment:
If we are adding this, I think you need to handle this in the cancellation message to not delete the upgrade but just fast forward the sequence